src/dotty/tools/dotc/typer/Inliner.scala

```diff
@@ -0,0 +1,200 @@
+package dotty.tools
+package dotc
+package typer
+
+import dotty.tools.dotc.ast.Trees.NamedArg
+import dotty.tools.dotc.ast.{Trees, untpd, tpd, TreeTypeMap}
+import Trees._
+import core._
+import Flags._
+import Symbols._
+import Types._
+import Decorators._
+import StdNames.nme
+import Contexts.Context
+import Names.Name
+import SymDenotations.SymDenotation
+import Annotations.Annotation
+import transform.ExplicitOuter
+import config.Printers.inlining
+import ErrorReporting.errorTree
+import util.Attachment
+import collection.mutable
+
+object Inliner {
+  import tpd._
+
+  private class InlinedBody(tree: => Tree) {
+    lazy val body = tree
+  }
+
+  private val InlinedBody = new Attachment.Key[InlinedBody]
+
+  def attachBody(inlineAnnot: Annotation, tree: => Tree)(implicit ctx: Context): Unit =
+    inlineAnnot.tree.putAttachment(InlinedBody, new InlinedBody(tree))
+
+  def inlinedBody(sym: SymDenotation)(implicit ctx: Context): Tree =
+    sym.getAnnotation(defn.InlineAnnot).get.tree
+      .attachment(InlinedBody).body
+
+  private class Typer extends ReTyper {
```

Comment:
1. inlining needs to be done during typing, since macro expansion depends on it, and macro expansion needs to e.g. create implicit values.

2. The inliner could possibly be formulated as another TreeMap instead of a full typer. I made it a type primarily for two reasons.

  2.1 That way we get constant folding for free (would have to be implemented explicitly in a treemap).
  2.2 That way we make sure the inlined code is type correct instead of failing during translation.